### PR TITLE
feat: add `@sapphire/lexure` package

### PR DESCRIPTION
**Pull request by kyranet**
## TODO

- [ ] Docs
- [x] Tests

### CODE REVIEW

**Comment by c43721** [comment]
omg omg omg omg

I can write the docs at a later point (more like help with docs)

File: packages/lexure/src/lib/lexer/streams/raw/TokenStream.ts [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+	private getParameter(): WordToken {
+		const index = this.input.indexOf(this.separator, this.position);
+		const value = index === -1 ? this.input.slice(this.position) : this.input.slice(this.position, index);
+		this.position += value.length;
+		return { type: TokenType.Parameter, value };
+	}
+}
+
+export const enum TokenType {
```
Comment (vladfrangu) [inline packages/lexure/src/lib/lexer/streams/raw/TokenStream.ts:58]: If this is exported to end users, dont make it const

File: packages/lexure/src/lib/parser/Parser.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
+import type { Parameter } from '../lexer/streams/ParameterStream';
+import type { IUnorderedStrategy } from './strategies/IUnorderedStrategy';
+import { ParserResult } from './ParserResult';
+import { EmptyStrategy } from './strategies/EmptyStrategy';
+
+export class Parser {
+	public strategy: IUnorderedStrategy;
+
+	public constructor(strategy: IUnorderedStrategy) {
```
Comment (UndiedGamer) [inline packages/lexure/src/lib/parser/Parser.ts:9]:
```suggestion
	public constructor(strategy?: IUnorderedStrategy) {
```
This is to make the next line valid

**Comment by favna** [comment]
PR still has TODOs so converted to draft

File: packages/lexure/package.json [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+	"publishConfig": {
+		"access": "public"
+	},
+	"dependencies": {
+		"@sapphire/result": "workspace:^"
+	},
+	"devDependencies": {
+		"tsup": "^6.1.3",
+		"typescript": "^4.7.4"
+	}
```
Comment (favna) [inline packages/lexure/package.json:65]:
```suggestion
	"devDependencies": {
		"@favware/cliff-jumper": "^1.8.5",
		"tsup": "^6.1.3",
		"typescript": "^4.7.4"
	}
```

**Review by favna — CHANGES_REQUESTED** [review]

File: packages/lexure/src/lib/parser/Parser.ts [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+		this.strategy = strategy ?? new EmptyStrategy();
+	}
+
+	public setUnorderedStrategy(strategy: IUnorderedStrategy) {
+		this.strategy = strategy;
+		return this;
+	}
+
+	public run(input: Iterable<Parameter>) {
+		return new ParserResult(this).parse(input);
```
[the remark this answers is not in the document]
Comment (UndiedGamer) [reply packages/lexure/src/lib/parser/Parser.ts:19]: mfw this has been in pending since 3 days

**Review by UndiedGamer — CHANGES_REQUESTED** [review]

File: packages/lexure/src/lib/parser/Parser.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (UndiedGamer) [reply packages/lexure/src/lib/parser/Parser.ts:19]: This was resolved in a different way

**Review by c43721 — DISMISSED** [review]

File: packages/lexure/README.md [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+A powerful and fast parser and utilities for non-technical user input, inspired by [`lexure`](https://www.npmjs.com/package/lexure), powered by [`@sapphire/result`](https://www.npmjs.com/package/@sapphire/result).
+
+The code in this package has been greatly inspired by [lexure] from [1Computer1].
+
+## Features
+
+-   Written in TypeScript
+-   Bundled with esbuild so it can be used in NodeJS and browsers
+-   Offers CommonJS, ESM and UMD bundles
+-   ~~Fully tested~~ WIP
```
Comment (favna) [inline packages/lexure/README.md:27]:
```suggestion
-   Fully tested
```

Done enough for this claim

**Review by favna — CHANGES_REQUESTED** [review]

**Review by favna — DISMISSED** [review]

**Review by c43721 — DISMISSED** [review]

File: packages/lexure/src/lib/lexer/Lexer.ts [hunk 6]
```diff
@@ -0,0 +1,27 @@
+import { ParameterStream } from './streams/ParameterStream';
+import { TokenStream } from './streams/raw/TokenStream';
+
+export class Lexer {
+	public readonly quotes: readonly [string, string][];
```
Comment (vladfrangu) [inline packages/lexure/src/lib/lexer/Lexer.ts:5]:
```suggestion
	public readonly quotes: readonly [open: string, close: string][];
```

File: packages/lexure/src/lib/lexer/Lexer.ts [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+	public raw(input: string) {
+		return new TokenStream(this, input);
+	}
+}
+
+export namespace Lexer {
+	export interface Options {
+		separator?: string;
+		quotes?: readonly [string, string][];
```
Comment (vladfrangu) [inline packages/lexure/src/lib/lexer/Lexer.ts:25]:
```suggestion
		quotes?: readonly [open: string, close: string][];
```

**Review by vladfrangu — CHANGES_REQUESTED** [review]

**Review by vladfrangu — APPROVED** [review]

**Review by c43721 — APPROVED** [review]

**Review by favna — APPROVED** [review]

**Pull request merged**